File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) == 0:
            return True
        
        stack = []
        
        for char in s:
            if char in "({[":
                stack.append(char)
            else:
                if not stack:
                    return False
                elif not self.isMatching(stack[-1], char):
                    return False
                else:
                    stack.pop()
        
        return len(stack) == 0
    
    def isMatching(self, a: str, b: str) -> bool:
        return (a == '(' and b == ')') or \
               (a == '{' and b == '}') or \
               (a == '[' and b == ']')

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        if(s.length() == 0){
            return true;
        }
        Stack <Character> stack = new Stack<>();
        for(int i=0;i<s.length();i++){
            char x = s.charAt(i);
            if(x=='(' || x=='{' || x=='['){
                stack.push(x);
            }else{
                if(stack.isEmpty()==true)
                return false;
                else if(isMatching(stack.peek(),x) == false)
                return false;
                else
                stack.pop();
            }
        } 
           return (stack.isEmpty()==true);
    }
        public boolean isMatching(char a,char b){
            return ((a=='(' && b==')') || (a=='{' && b=='}') || (a=='[' && b==']'));
        }
}